core/src/main/scala/cats/data/Chain.scala

```diff
@@ -120,13 +123,19 @@ sealed abstract class Chain[+A] {
   /**
    * Returns true if there are no elements in this collection.
    */
-  def isEmpty: Boolean = !this.isInstanceOf[Chain.NonEmpty[_]]
+  def isEmpty: Boolean = this eq Chain.Empty
```

Comment:
I have got a feeling that `eq` should work faster than `isInstanceOf`...

Reply:
@johnynek I realized that you defined this method via `isInstanceOf` in #3533.
Not I am in doubt... Is `isIntsanceOf` more performant than just `eq`, could you clarify please?

Reply:
Oh.. Seems I've caught it (perhaps). I've checked the bytecode and realized that the
`!this.isInstanceOf[Chain.NonEmpty[_]]` trick lets to avoid a `getstatic` instruction compared to `eq` which still requires it. 

Is that what you meant here?

Reply:
Rolled it back.